### ISSUE TITLE
Remove special case for `invariant`

### ISSUE THREAD

**Issue by petehunt** [facebook/flow#112]
I'm finding it difficult to add type annotations to my function named `invariant()` because of this little guy: https://github.com/facebook/flow/blob/0f56d1742967b855ab5ee5eaee93d2edfc503e54/src/typing/type_inference_js.ml#L2058

It seems that many of these functions are special cased. I'm curious as to why, because I can't use any functions that share these names anymore.

**Comment by samwgoldman** [comment]
The proposal for refinement carrying boolean types in #34 isn't powerful enough, AFAICT, to express the type of `invariant`. Since TypeScript has already implemented that, it's probably a good time to start thinking about how we'd express something like `invariant` in the type language.

**Comment by leebyron** [comment]
Here's a rough working but not long-term stable proposal:

```js
invariant(predicate: any): $Asserts<predicate is true>
```

This uses the `Ident "is" Type` grammar proposed in #34. It uses @mroch's recent addition of boolean literals as Types. And it uses Flow's "this is an uncommon and undocumented" pattern of $Thing for communicating specific typing behavior.

**Comment by samwgoldman** [comment]
I like this syntax and I think the syntax does communicate all the necessary information. I feel like the implementation might get tricky around the `$Assert` type, which should result in an [abnormal event](https://github.com/facebook/flow/blob/ac47977acedc6d7e6e696bc63ad69032b85842a0/src/typing/type_inference_js.ml#L29). I haven't looked into this enough yet to say whether it's realistic or not (maybe @avikchaudhuri can weigh in), but I'll give this some more thought.

**Comment by mroch** [comment]
3e35bd4 should get the rest of the static_upstream/core special cases out of the way, except for `invariant`. Renamed this task to preserve the discussion of predicate types.

**Comment by gajus** [comment]
>     invariant(predicate: any): $Asserts<predicate is true>

Was this ever implemented?

**Comment by leebyron** [comment]
No, `%checks` was built which is for predicate functions but not assertion functions. If this is a model the flow team is happy with then maybe something like this could be built such as `%asserts`?

**Comment by gajus** [comment]
For others reference, here is the documentation for `%checks` https://flow.org/en/docs/types/functions/#toc-predicate-functions.

**Comment by gajus** [comment]
This issue is duplicate of https://github.com/facebook/flow/issues/34.

**Comment by mattconde** [comment]
So I think I've caught up on this issue, am I right to think I still can't flowtype my own "invariant" module correctly because of this?

**Comment by goodmind** [comment]
TypeScript now has this https://github.com/microsoft/TypeScript/pull/32695
